src/native/libs/System.Native/pal_uid.c

```diff
@@ -237,3 +238,104 @@ int32_t SystemNative_GetGroups(int32_t ngroups, uint32_t* groups)
 
     return getgroups(ngroups, groups);
 }
+
+char* SystemNative_GetGroupName(uint32_t gid)
+{
+#if HAVE_GETGRGID_R
+    size_t bufferLength = 512;
+    while (1)
+    {
+        char *buffer = (char*)malloc(bufferLength);
+        if (buffer == NULL)
+        {
+            return NULL;
+        }
+
+        struct group* result;
+        struct group gr;
+        if (getgrgid_r(gid, &gr, buffer, bufferLength, &result) == 0)
+        {
+            if (result == NULL)
+            {
+                errno = ENOENT;
+                free(buffer);
+                return NULL;
+            }
+            else
+            {
+                char* name = strdup(gr.gr_name);
+                free(buffer);
+                return name;
+            }
+        }
+
+        free(buffer);
+        size_t tmpBufferLength;
+        if (errno != ERANGE || !multiply_s(bufferLength, (size_t)2, &tmpBufferLength))
+        {
+            return NULL;
+        }
+        bufferLength = tmpBufferLength;
+    }
+#else
+    // Platforms like Android API level < 24 do not have getgrgid_r available
+    // Android API level >= 24 do have getgrgid_r available but call getgrgid internally anyway
+    struct group* result = getgrgid(gid);
+    if (result == NULL)
+    {
+        return NULL;
+    }
+    char* name = strdup(result->gr_name);
+    return name;
+#endif
+}
+
+char* SystemNative_GetUserName(uint32_t uid)
+{
+#if HAVE_GETPWUID_R
+    size_t bufferLength = 512;
+    while (1)
+    {
+        char *buffer = (char*)malloc(bufferLength);
+        if (buffer == NULL)
+        {
+            return NULL;
+        }
+
+        struct passwd pw;
+        struct passwd* result;
+        if (getpwuid_r(uid, &pw, buffer, bufferLength, &result) == 0)
+        {
+            if (result == NULL)
+            {
+                errno = ENOENT;
+                free(buffer);
+                return NULL;
+            }
+            else
+            {
+                char* name = strdup(pw.pw_name);
+                free(buffer);
+                return name;
+            }
+        }
+
+        free(buffer);
+        size_t tmpBufferLength;
+        if (errno != ERANGE || !multiply_s(bufferLength, (size_t)2, &tmpBufferLength))
+        {
+            return NULL;
+        }
+        bufferLength = tmpBufferLength;
+    }
+#else
+    // Platforms like tvOS/iOS might not have getpwuid_r available.
```

Comment:
I just checked and I see HAVE_GETPWUID_R as defined for iOS/tvOS, so you're probably hitting some other issue

Reply:
so looks like we explicitly disabled `getpwuid_r` in mono/mono, I just tried and while it exists it always returns `result == NULL`, same for `getpwuid` so we'll need to make do without it on these platforms.

Reply:
Btw. some of that code is duplicated with `SystemNative_GetPwUidR`

Reply:
I had a typo. The `#if` should have been either `#ifdef` or `#if defined`.

But I'll also address what you mentioned about both `getpwuid_r` and `getpwuid` being useless in ios/tvos because they both return `NULL`.

Reply:
So besides being disabled [for IOS in mono](https://github.com/dotnet/runtime/blob/7414af2a5f6d8d99efc27d3f5ef7a394e0b23c42/src/mono/cmake/configure.cmake#L188-L190), it is [also disabled in HOST_WASI](https://github.com/dotnet/runtime/blob/e4ac5cd051107772b99cf3f275f2718ff20120a4/src/mono/cmake/configure.cmake#L268-L276). Am I right to assume that's WASM?

Reply:
> I had a typo. The #if should have been either #ifdef or #if defined.

I was wrong, it wasn't a typo. In `pal_config.h.in` I used `#cmakedefine01 HAVE_GETPWUID_R` which means it will always be defined, with either 0 or 1 as value.

So using `#if VALUE` or `#if !VALUE` was correct in this case.

Reply:
> So besides being disabled [for IOS in mono](https://github.com/dotnet/runtime/blob/7414af2a5f6d8d99efc27d3f5ef7a394e0b23c42/src/mono/cmake/configure.cmake#L188-L190), it is [also disabled in HOST_WASI](https://github.com/dotnet/runtime/blob/e4ac5cd051107772b99cf3f275f2718ff20120a4/src/mono/cmake/configure.cmake#L268-L276). Am I right to assume that's WASM? I suspect the answer [is yes](https://github.com/WebAssembly/wasi-sdk).

I was originally planning to check for the current target platform to determine if I had to use a mutex or not, but decided not to, because I'm hoping the methods get fixed eventually on the ios/tvos side, so we better have the code working on our side.

The one thing left for me to address is checking which tests will fail in ios/tvos with the latest commit, and remove the group/user checks in those two platforms.